File: sims/dashboard/lib/diff.ts
import type { DiffLine } from "./types";

// Marker order matters: +++/--- must be classified as meta before the +/- checks.
export function classifyDiff(raw: string): DiffLine[] {
  if (!raw.trim()) return [];
  return raw.split("\n").map((text) => {
    let cls: DiffLine["cls"] = "ctx";
    if (text.startsWith("diff ") || text.startsWith("index ") || text.startsWith("+++") || text.startsWith("---")) cls = "meta";
    else if (text.startsWith("@@")) cls = "hunk";
    else if (text.startsWith("+")) cls = "add";
    else if (text.startsWith("-")) cls = "del";
    return { cls, text };
  });
}

File: eval-harness/backend/internal/artifacts/diff.go
package artifacts

import "strings"

// ClassifyDiff classifies each line of a unified diff by its leading marker.
// Marker order matters: +++/--- are classified as meta before the +/- checks.
func ClassifyDiff(raw string) []DiffLine {
	if strings.TrimSpace(raw) == "" {
		return []DiffLine{}
	}
	lines := strings.Split(raw, "\n")
	out := make([]DiffLine, len(lines))
	for i, text := range lines {
		cls := "ctx"
		switch {
		case strings.HasPrefix(text, "diff "),
			strings.HasPrefix(text, "index "),
			strings.HasPrefix(text, "+++"),
			strings.HasPrefix(text, "---"):
			cls = "meta"
		case strings.HasPrefix(text, "@@"):
			cls = "hunk"
		case strings.HasPrefix(text, "+"):
			cls = "add"
		case strings.HasPrefix(text, "-"):
			cls = "del"
		}
		out[i] = DiffLine{Cls: cls, Text: text}
	}
	return out
}
